src/test/mod.rs

```diff
@@ -217,10 +220,143 @@ fn transaction_rollback() {
+            let ret_inner =
+                self::test::dsl::test.load::<(Option<i64>, Option<String>, Option<i64>)>(&conn)?;
+            assert_eq!(ret_inner.len(), 2);
+            Ok(())
+        });
+        assert_result!(out_inner);
+        let ret = self::test::dsl::test.load::<(Option<i64>, Option<String>, Option<i64>)>(&conn)?;
+        assert_eq!(ret.len(), 2);
+        Ok(())
+    });
+    assert_result!(out);
+    let ret = self::test::dsl::test.load::<(Option<i64>, Option<String>, Option<i64>)>(&conn);
+    assert_result!(ret);
+    assert_eq!(ret.unwrap().len(), 2);
+}
+
+#[test]
+fn transaction_nested_commit_rollback() {
+    let conn = init_testing();
+
+    clean_test(&conn);
+
+    let ret = conn.execute(CREATE_TEST_TABLE);
+    assert_result!(ret);
+    let out = conn.transaction::<_, Error, _>(|| {
+        let sql = format!("INSERT INTO test ({}) VALUES ({})", "TST_CHR", TEST_VARCHAR);
+        let _ret = conn.execute(&*sql)?;
+        let ret = self::test::dsl::test.load::<(Option<i64>, Option<String>, Option<i64>)>(&conn)?;
+        assert_eq!(ret.len(), 1);
+
+        let out_inner = conn.transaction::<i32, Error, _>(|| {
+            let sql_inner = format!("INSERT INTO test ({}) VALUES ({})", "TST_CHR", TEST_VARCHAR);
+            let _ret_inner = conn.execute(&*sql_inner)?;
+            let ret_inner =
+                self::test::dsl::test.load::<(Option<i64>, Option<String>, Option<i64>)>(&conn)?;
+            assert_eq!(ret_inner.len(), 2);
+            Err(Error::NotFound)
+        });
+        assert!(out_inner.is_err() && !out_inner.is_ok(), "What :shrug:?");
+        let ret = self::test::dsl::test.load::<(Option<i64>, Option<String>, Option<i64>)>(&conn)?;
+        assert_eq!(ret.len(), 1);
+        Ok(())
+    });
+    assert_result!(out);
+    let ret = self::test::dsl::test.load::<(Option<i64>, Option<String>, Option<i64>)>(&conn);
+    assert_result!(ret);
+    assert_eq!(ret.unwrap().len(), 1);
+}
+
+#[test]
+fn transaction_nested_rollback_commit() {
+    let conn = init_testing();
+
+    clean_test(&conn);
+
+    let ret = conn.execute(CREATE_TEST_TABLE);
+    assert_result!(ret);
+    let out = conn.transaction::<i32, Error, _>(|| {
+        let sql = format!("INSERT INTO test ({}) VALUES ({})", "TST_CHR", TEST_VARCHAR);
+        let _ret = conn.execute(&*sql)?;
+        let ret = self::test::dsl::test.load::<(Option<i64>, Option<String>, Option<i64>)>(&conn)?;
+        assert_eq!(ret.len(), 1);
+
+        let out_inner = conn.transaction::<_, Error, _>(|| {
+            let sql_inner = format!("INSERT INTO test ({}) VALUES ({})", "TST_CHR", TEST_VARCHAR);
+            let _ret_inner = conn.execute(&*sql_inner)?;
+            let ret_inner =
+                self::test::dsl::test.load::<(Option<i64>, Option<String>, Option<i64>)>(&conn)?;
+            assert_eq!(ret_inner.len(), 2);
+            Ok(())
+        });
+        assert_result!(out_inner);
```

Comment:
There should be a second inner transaction that is rolled back.

Reply:
We insert something in the first transaction `a`. Then we open another transaction `b` and insert another value. After that we commit `b`. So `a` sees the change. Then we rollback `a` with an arbitrary error. The initial connection doesn't see anything of neither `a` nor `b`. Why would we need another inner transaction `c`? Should that go in `a` or in `b`?

Reply:
We need a second inner transaction because we know about the inner details of how nested transactions are implemented. The outer level uses the oci level transaction, whereas the inner transaction uses savepoints. So we need to check that both rollback variants are working.

Reply:
I see what you mean. I think we should wrap this in new tests.